elasticsearch/connection/http_urllib3.py

```diff
@@ -62,33 +76,37 @@ def __init__(self, host='localhost', port=9200, http_auth=None,
                 self.headers[k.lower()] = headers[k]
 
         self.headers.setdefault('content-type', 'application/json')
-        ca_certs = CA_CERTS if ca_certs is None else ca_certs
         pool_class = urllib3.HTTPConnectionPool
         kw = {}
         if use_ssl:
+            if not ca_certs and not ssl_context and verify_certs:
+                # If no ca_certs and no sslcontext passed and asking to verify certs
+                # raise error
+                raise ImproperlyConfigured("Root certificates are missing for certificate "
+                    "validation. Either pass them in using the ca_certs parameter or "
+                    "install certifi to use it automatically.")
+
+            if not ssl_context:
+                # if SSLContext hasn't been passed in, create one.
+                cafile= CA_CERTS if ca_certs is None else ca_certs
+                ssl_context = create_ssl_context(cafile=cafile)
+
             pool_class = urllib3.HTTPSConnectionPool
+
+            if not verify_certs:
+                ssl_context.check_hostname = False
+                ssl_context.verify_mode = ssl.CERT_NONE
+                warnings.warn(
+                    'Connecting to %s using SSL with verify_certs=False is insecure.' % host)
+
             kw.update({
                 'ssl_version': ssl_version,
                 'assert_hostname': ssl_assert_hostname,
                 'assert_fingerprint': ssl_assert_fingerprint,
+                'ssl_context': ssl_context,
+                'cert_file': client_cert,
+                'key_file': client_key,
```

Comment:
Shouldn't `cert_file` and `key_file` also be part of `ssl_context`? seems inconsistent to pass some values directly and some via ssl context...

Reply:
@fxdgear @HonzaKral is right. If you have ssl_context you do not need cert_file and key_file. But it is depend on the python version.
If someone using python 2.7.8 ( ssl_context is introduced in 2.7.9!), can not use ssl_context. So, cert_file and key_file is used.
For example:
python 2.7.8:
```python
import ssl 
ssl.wrap_socket(socket, keyfile, certfile, ca_certs, cert_reqs=ssl.CERT_REQUIRED)
```
python 2.7.9:
```python
import ssl
context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
#load the certificates
....
context.wrap_socket(socket, serverside=False)
```
I propose to remove cert_file and key_file, because it is inconsistent and up to the developer to create the ssl_context. We are moved to python 2.7.13 this solution is okay for me. But when I opened #471 is was not the case!